src/Drivers/nlpPriDec_ex9_user_recourse.hpp

```diff
@@ -0,0 +1,475 @@
+      } else {
+        assert(con_idx==m-1);
+        cons[m-1] = (1-x[0]+xi_[0])*(1-x[0]+xi_[0]);
+        for(int i=1;i<nS_;i++) {
+          cons[m-1] += (x[i] + xi_[i])*(x[i] + xi_[i]);
+        }
+        for(int i=nS_;i<nx_;i++) {
+          cons[m-1] += x[i]*x[i];
+        }
+      }
+      /*if(idx_<=10 && con_idx<100) {
+        cons[con_idx] = x[con_idx]-x_[idx];
+      }*/    
+
+    }
+    return true; 
+  }
+  
+  //  r_i(x;\xi^i) = 1/S *  min_y 0.5 || y - x ||^2 such that 
+  bool eval_grad_f(const long long& n, const double* x, bool new_x, double* gradf)
+  {
+    assert(ny_==n);    
+    for(int i=0;i<nx_;i++) gradf[i] = (x[i]-x_[i])/double(S_);
+
+    return true;
+  }
+ 
+  virtual bool eval_Jac_cons(const long long& n, 
+                             const long long& m, 
+                             const long long& num_cons, 
+                             const long long* idx_cons,
+                             const double* x, 
+                             bool new_x,
+                             const long long& nsparse, 
+                             const long long& ndense, 
+                             const int& nnzJacS, 
+                             int* iJacS, 
+                             int* jJacS, 
+                             double* MJacS, 
+                             double* JacD)
+  {
+    assert(num_cons==nx_||num_cons==0);
+    //indexes for sparse part
+    if(num_cons==0) {
+      return true;
+    }
+
+    if(iJacS!=NULL && jJacS!=NULL) {
+      int nnzit=0;
+      for(int itrow=0; itrow<num_cons; itrow++) {
+        const int con_idx = (int) idx_cons[itrow];
+        if(con_idx<nsparse_-1) {
+          //sparse Jacobian eq w.r.t. x and s
+          //yk
+          iJacS[nnzit] = con_idx;
+          jJacS[nnzit] = con_idx; //-1
+          nnzit++;
+
+          //yk+1
+          iJacS[nnzit] = con_idx;
+          jJacS[nnzit] = con_idx+1; //1
+          nnzit++;
+        } else if (con_idx==nsparse_-1) {
+          iJacS[nnzit] = con_idx;
+          jJacS[nnzit] = con_idx; //-1
+          nnzit++;
+        } else if (con_idx==m-1) { 
+          iJacS[nnzit] = m-1;
+          jJacS[nnzit] = 0;
+          nnzit++;
+          //cons[m-1] = (1-x[0]+xi_[0]);
+          if(nsparse_<=nS_) {
+            for(int i=1;i<nsparse_;i++) {
+              iJacS[nnzit] = m-1;
+              jJacS[nnzit] = i;
+              nnzit++;
+              //cons[m-1] += (x[i] + xi_[i])*(x[i] + xi_[i]);
+            }
+          } else {
+            for(int i=1;i<nS_;i++) {
+              iJacS[nnzit] = m-1;
+              jJacS[nnzit] = i;
+              nnzit++;
+              //cons[m-1] += (x[i] + xi_[i])*(x[i] + xi_[i]);
+            }
+            for(int i=nS_;i<nsparse_;i++) {
+              iJacS[nnzit] = m-1;
+              jJacS[nnzit] = i;
+              nnzit++;
+              //cons[m-1] += x[i]*x[i];
+            }
+          }
+          //sparse Jacobian ineq w.r.t x and s
+        }
+      }
```

Comment:
Consider rewriting this function in a thread-safe way. That will simplify OpenMP or CUDA implementations.

Reply:
Hoping to discuss this in detail later this week.

Reply:
To be addressed in a separate PR.